pkg/logutil/zap_log_test.go

```diff
@@ -0,0 +1,229 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net"
+	"os"
+	"time"
+	"unsafe"
+
+	. "github.com/pingcap/check"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var _ = Suite(&testZapLogSuite{})
+
+type testZapLogSuite struct{}
+
+// testingWriter is a WriteSyncer that writes to the the messages.
+type testingWriter struct {
+	c        *C
+	messages []string
+}
+
+func newTestingWriter(c *C) *testingWriter {
+	return &testingWriter{c: c}
+}
+
+func (w *testingWriter) Write(p []byte) (n int, err error) {
+	n = len(p)
+	p = bytes.TrimRight(p, "\n")
+	m := fmt.Sprintf("%s", p)
+	w.messages = append(w.messages, m)
+	return n, nil
+}
+
+func (w *testingWriter) Sync() error {
+	return nil
+}
+
+type verifyLogger struct {
+	*zap.Logger
+	w *testingWriter
+}
+
+func (v *verifyLogger) AssertMessage(msg ...string) {
+	for i, m := range msg {
+		v.w.c.Assert(m, Equals, v.w.messages[i])
+	}
+}
+
+func newZapTestLogger(cfg *LogConfig, c *C) verifyLogger {
+	writer := newTestingWriter(c)
+	opt := cfg.buildOptions(writer)
+	level := zap.NewAtomicLevel()
+	err := level.UnmarshalText([]byte(cfg.Level))
+	c.Assert(err, IsNil)
+	lg := zap.New(zapcore.NewCore(newZapTextEncoder(cfg), writer, level), opt...)
+	return verifyLogger{
+		Logger: lg,
+		w:      writer,
+	}
+}
+
+type username string
+
+func (n username) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddString("username", string(n))
+	return nil
+}
+
+func (t *testZapLogSuite) TestLog(c *C) {
+	conf := &LogConfig{Level: "debug", File: FileLogConfig{}, DisableTimestamp: true}
+	lg := newZapTestLogger(conf, c)
+	sugar := lg.Sugar()
+	defer sugar.Sync()
+	sugar.Infow("failed to fetch URL",
+		"url", "http://example.com",
+		"attempt", 3,
+		"backoff", time.Second,
+	)
+	sugar.Infof(`failed to "fetch" [URL]: %s`, "http://example.com")
+	sugar.Debugw("Slow query",
+		"sql", `SELECT * FROM TABLE
+	WHERE ID="abc"`,
+		"duration", 1300*time.Millisecond,
+		"process keys", 1500,
+	)
+	sugar.Info("Welcome")
+	sugar.Info("Welcome TiDB")
+	sugar.Info("欢迎")
+	sugar.Info("欢迎来到 TiDB")
+	sugar.Warnw("Type",
+		"Counter", math.NaN(),
+		"Score", math.Inf(1),
+	)
+	lg.Info("Testing typs",
+		zap.String("filed1", "noquote"),
+		zap.String("filed2", "in quote"),
+		zap.Strings("urls", []string{"http://mock1.com:2347", "http://mock2.com:2432"}),
+		zap.Strings("urls-peer", []string{"t1", "t2 fine"}),
+		zap.Uint64s("store ids", []uint64{1, 4, 5}),
+		zap.Object("object", username("user1")),
+		zap.Object("object2", username("user 2")),
+		zap.Binary("binary", []byte("ab123")),
+		zap.Bool("is processed", true),
+		zap.ByteString("bytestring", []byte("noquote")),
+		zap.ByteString("bytestring", []byte("in quote")),
+		zap.Int8("int8", int8(1)),
+		zap.Uintptr("ptr", 0xa),
+		zap.Reflect("reflect", []int{1, 2}),
```

Comment:
The test is `lg.Info("Test", zap.Reflect("reflect111", []string{"r1", "r2"}))`, and the result is `[INFO] [zap_log_test.go:113] [Test] [reflect111="[\"r1\",\"r2\"]"]`. 
Is this result in line with expectations?

Reply:
I think it's expected. RFC does not care about the how object encoded, it cares the encoded should be a string and if have any need escape chart in that should escape the encoded string. maybe this is different from `zap.Strings()`, but I think this method will rarely be used.